### PR TITLE
fix: cert rendering temporary

### LINKED ISSUE #42893: Claiming Certificate from Superblock
This issue outlines _quality-of-life_ improvements to the workflow of claiming a certificate from the superblock.

**Issues**
1) [ ] Navigating to a superblock temporarily renders the ClaimCertSteps card, despite the certificate having been claimed. [1]
2) [ ] After claiming a certificate from superblock, flash message obscures _Show Certification_ button (margin to move button down?)
3) [ ] A signed-out Camper should probably not see the same card [2]

**Foot notes**
1.

https://user-images.githubusercontent.com/51722130/125916460-29fae27a-179f-49f4-b058-80dd4e5d74b5.mp4

2.
image

### PR DESCRIPTION
**Pull request by Sembauke**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #42893

### CODE REVIEW

**Comment by Sembauke** [comment]
Let me know what you think about the extra step for claiming a certification.

**Review by ShaunSHamilton — DISMISSED** [review]
Thanks, for tackling this, Sem.

I would prefer not bundling state like `isSignedIn` with the `claimCertSteps`. Would you mind using the selector in the component dispatch instead, and passing it through as a prop?

**Comment by Sembauke** [comment]
> Thanks, for tackling this, Sem.
> 
> I would prefer not bundling state like `isSignedIn` with the `claimCertSteps`. Would you mind using the selector in the component dispatch instead, and passing it through as a prop?

This would mean that `isSignedIn` will not count as an extra step. So instead of `0/6`  it would be `0/5`

**Comment by ShaunSHamilton** [comment]
> > Thanks, for tackling this, Sem.
> > I would prefer not bundling state like `isSignedIn` with the `claimCertSteps`. Would you mind using the selector in the component dispatch instead, and passing it through as a prop?
> 
> This would mean that `isSignedIn` will not count as an extra step. So instead of `0/6` it would be `0/5`

Correct. Campers who are not signed in should not see the steps card, as per the linked issue.

**Comment by Ewelin36** [comment]
Dne ne 29. 8. 2021 13:05 uživatel github-actions[bot] <
***@***.***> napsal:

File: client/i18n/locales/english/translations.json [hunk 1]
```diff
@@ -567,6 +567,7 @@
     "title": "Claim Your Certification",
     "intro": "Complete the following steps to claim and view your {{i18nCertText}}",
     "complete-project": "Complete {{i18nCertText}} Projects",
+    "login": "Login to claim certification",
```
Comment (ShaunSHamilton) [inline client/i18n/locales/english/translations.json:570]: If you could revert this change

File: client/src/templates/Introduction/components/CertChallenge.js [hunk 2]
```diff
@@ -82,6 +102,10 @@ const CertChallenge = ({
   const { certSlug } = certMap.find(x => x.title === title);
 
   useEffect(() => {
+    if (complete && !pending) {
+      setUserLoaded(true);
+    }
```
Comment (ShaunSHamilton) [inline client/src/templates/Introduction/components/CertChallenge.js:107]: Take this with a grain of salt:

As `complete` and `pending` are props, we should be adding them to the dependency array of **their own** `useEffect`. Also, I do not think they should be _accessed_ (destructured) within the root of the function - should be destructured within the `useEffect` callback.

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Once this is changed, all LGTM. Thanks, Sem.

File: client/src/templates/Introduction/components/CertChallenge.js (hunk 2 above)
Comment (Sembauke) [reply client/src/templates/Introduction/components/CertChallenge.js:107]: Thanks for the review 👍

File: client/src/templates/Introduction/components/CertChallenge.js [hunk 3]
```diff
@@ -118,10 +145,9 @@ const CertChallenge = ({
       ? verifyCert(certSlug)
       : createFlashMessage(honestyInfoMessage);
   };
-
   return (
     <div className='block'>
-      {(!isCertified || !canViewCert) && (
+      {!isCertified && userLoaded && isSignedIn && (
```
Comment (ShaunSHamilton) [inline client/src/templates/Introduction/components/CertChallenge.js:150]: Correct me, if I am wrong, but this still needs to hold for if a Camper has not accepted the academic honesty etc.
```suggestion
      {(!isCertified || !canViewCert) && userLoaded && isSignedIn && (
```

Comment (Sembauke) [reply client/src/templates/Introduction/components/CertChallenge.js]: I hardly could tell the difference between isCertified and canViewCert

**Comment by Sembauke** [comment]
I don't know why but the cert temp renders again because of this change.

File: client/src/templates/Introduction/components/CertChallenge.js (hunk 3 above)
Comment (ShaunSHamilton) [reply client/src/templates/Introduction/components/CertChallenge.js]: `isCertified` is when a Camper has claimed a certificate. `canViewCert` is for when a Camper has made the correct settings public to allow a certificate to be viewed.

These two could possibly be combined into one, but at some point we might want slightly different behaviour.

**Comment by Sembauke** [comment]
So I figured out what is going wrong:

The certification card renders on these conditions:
`userLoaded && isSignedIn && (!isCertified || !canViewCert)`

if a user is not loaded and is certified the following would happen:
` userLoaded: false, isSignedIn: true, isCertified: false,  canViewCert: false)` 

the condition would come out as false.

But when the user is loaded:
 ` userLoaded: true, isSignedIn: true, isCertified: true,  canViewCert: false)` 

the condition would come out as true, thus still temporarily rendering the certification.
This probably happens because the function that counts the steps is not done counting.

In the next render: 
` userLoaded: true, isSignedIn: true, isCertified: true,  canViewCert: true)` 

the condition would come out as false.

-----
I like to hear your thoughts.

**Comment by ojeytonwilliams** [comment]
@Sembauke could you expand a bit about what's going wrong?

**Comment by Sembauke** [comment]
@ojeytonwilliams 
At the moment the certification steps render for split seconds and then disappear again.  
The explanation above is what I found out.

If you want to take a look yourself go ahead.

Here a slow-motion video:
https://imgur.com/a/3EgXpgT

**Comment by ojeytonwilliams** [comment]
Thanks, Sem, I'm with you now.  How about we simply hide the Claim Your Certification component behind a skeleton until the user has loaded?  

It's a bit out of scope for this PR, but if they're not signed in at all, I think it makes sense to have that section be 

> Claim Your Certification
> Sign in to save your progress and claim certifications

**Comment by Sembauke** [comment]
> Thanks, Sem, I'm with you now. How about we simply hide the Claim Your Certification component behind a skeleton until the user has loaded?

I do not think that is possible with the currently implemented logic.

That is because the `canViewCert` is not relying on `isSigned` (it is not part of the steps that are asynchronously received).

A solution might be to set `canViewCert` to true if `isCertified` is true, thus not rendering the certification card.

I hope this makes sense.

**Comment by ojeytonwilliams** [comment]
I'm probably missing something, but my understanding is that the client can't know exactly what to render until the user object has returned from the api.  Is that correct?

If it is, we should change the render logic so that it shows a skeleton until the user has returned.

**Comment by Sembauke** [comment]
@ojeytonwilliams I am not sure but I think I fixed it, can you check and see if you like the new behavior?

**Comment by ShaunSHamilton** [comment]
This is confusing. I am getting that behaviour again where clicking close on the Donation Modal causes the viewport to jump to the top of the page.

File: client/src/templates/Introduction/components/CertChallenge.js [hunk 4]
```diff
@@ -118,10 +144,9 @@ const CertChallenge = ({
       ? verifyCert(certSlug)
       : createFlashMessage(honestyInfoMessage);
   };
-
   return (
     <div className='block'>
-      {(!isCertified || !canViewCert) && (
+      {userLoaded && isSignedIn && !isCertified && !canViewCert && (
```
Comment (ShaunSHamilton) [inline client/src/templates/Introduction/components/CertChallenge.js:149]:
```suggestion
      {userLoaded && isSignedIn && (!isCertified || !canViewCert) && (
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
I **_THINK_** this is what we need

File: client/src/templates/Introduction/components/CertChallenge.js (hunk 4 above)
Comment (ShaunSHamilton) [reply client/src/templates/Introduction/components/CertChallenge.js:149]: This was silly - I did not read above conversation.

**Review by ShaunSHamilton — DISMISSED** [review]
This LGTM 👍 . Nice job, Sem.

File: client/src/templates/Introduction/components/CertChallenge.js (hunk 4 above)
Comment (Sembauke) [reply client/src/templates/Introduction/components/CertChallenge.js]: No problem Shaun!

**Comment by ojeytonwilliams** [comment]
@Sembauke before we get this in, could you update the tests, since we know they don't catch the modal bug?

**Comment by ojeytonwilliams** [comment]
Thanks for updating the tests, Sem.  Just so I can QA this properly, which commit had the bug where it jumps to the top of the page?

**Review by ShaunSHamilton — DISMISSED** [review]
This makes sense with the button not showing to signed-out users. Thanks, Sem.

File: client/src/templates/Introduction/components/CertChallenge.js [hunk 5]
```diff
@@ -53,10 +69,14 @@ const CertChallenge = ({
...
   verifyCert,
   title,
+  fetchState,
+  isSignedIn,
   user: { isHonest, username }
 }) => {
   const [canClaim, setCanClaim] = useState({ status: false, result: '' });
   const [isCertified, setIsCertified] = useState(false);
+  const [userLoaded, setUserLoaded] = useState(false);
+  const [userChecked, setUserChecked] = useState(false);
```
Comment (ojeytonwilliams) [inline client/src/templates/Introduction/components/CertChallenge.js:79]: What does `userChecked` mean?  I think it would make more sense to have a `verificationFetchState` object to track `getVerifyCanClaimCert`.  Also, if you're feeling keen, `canClaim` really should be a boolean called `certIsClaimable` and `result` should be separate and called something meaningful. `verificationStatus`, maybe.  It's unclear to me what it's for.

@ShaunSHamilton might have more context.

File: client/src/templates/Introduction/components/CertChallenge.js [hunk 6]
```diff
@@ -118,27 +146,35 @@ const CertChallenge = ({
...
-        bsStyle='primary'
-        disabled={!canClaim.status || (isCertified && !canViewCert)}
-        href={certLocation}
-        onClick={createClickHandler(certSlug)}
-      >
-        {isCertified ? t('buttons.show-cert') : t('buttons.claim-cert')}
-      </Button>
+      {userLoaded &&
+        isSignedIn &&
+        (!isCertified || (!canViewCert && userChecked)) && (
```
Comment (ojeytonwilliams) [inline client/src/templates/Introduction/components/CertChallenge.js:153]:
```suggestion
      {userLoaded &&
        isSignedIn &&
        (!isCertified || (!canViewCert && userChecked)) && (
```
This is a bit complicated.  I'd recommend moving this into the render body and giving it a name.  `showCertificationCard` is probably fine.

File: client/src/templates/Introduction/components/CertChallenge.js [hunk 7]
```diff
@@ -96,7 +125,6 @@ const CertChallenge = ({
         stepVal => typeof stepVal === 'boolean' && stepVal
       ).length + projectsCompleted;
     const numberOfSteps = Object.keys(steps).length;
-
     setCanViewCert(completedCount === numberOfSteps);
```
Comment (ojeytonwilliams) [inline client/src/templates/Introduction/components/CertChallenge.js:128]:
```suggestion
    setCanViewCert(completedCount === numberOfSteps);
```
I know it's out of scope, but `canViewCert` is misleading.  It's really just `hasCompletedRequiredSteps`.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]

File: client/src/templates/Introduction/components/CertChallenge.js (hunk 5 above)
Comment (Sembauke) [reply client/src/templates/Introduction/components/CertChallenge.js]: Not sure what you mean by :

>  I think it would make more sense to have a verificationFetchState object to track getVerifyCanClaimCert

File: client/src/templates/Introduction/components/CertChallenge.js [hunk 8]
```diff
@@ -90,8 +94,8 @@ const CertChallenge = ({
         try {
           const data = await getVerifyCanClaimCert(username, superBlock);
           const { status, result } = data?.response?.message;
-          setCanClaim({ status, result });
-          setUserChecked(true);
+          setCertIsClaimable({ status, result });
+          SetVerificationFetchState(true);
```
Comment (ShaunSHamilton) [inline client/src/templates/Introduction/components/CertChallenge.js:98]:
```suggestion
          setVerificationFetchState(true);
```

File: client/src/templates/Introduction/components/CertChallenge.js [hunk 9]
```diff
@@ -73,15 +73,19 @@ const CertChallenge = ({
...
 }) => {
-  const [canClaim, setCanClaim] = useState({ status: false, result: '' });
+  const [certIsClaimable, setCertIsClaimable] = useState({
+    isClaimable: false,
+    isClaimableMessage: ''
+  });
   const [isCertified, setIsCertified] = useState(false);
   const [userLoaded, setUserLoaded] = useState(false);
-  const [userChecked, setUserChecked] = useState(false);
+  const [verificationFetchState, SetVerificationFetchState] = useState(false);
```
Comment (ShaunSHamilton) [inline client/src/templates/Introduction/components/CertChallenge.js:82]:
```suggestion
  const [verificationFetchState, setVerificationFetchState] = useState(false);
```

File: client/src/templates/Introduction/components/CertChallenge.js (hunk 5 above)
Comment (ojeytonwilliams) [reply client/src/templates/Introduction/components/CertChallenge.js]: So, we generally use an object

```
{   
  pending: boolean,
  complete: boolean,
  errored: boolean,
  error: null | string
}
```
to track the status of an async operation, so I was suggesting we do something similar here.  Essentially we perform an ajax operation and we conditionally display stuff on the page depending on what state that operation is in.

On reflection, it's probably over the top to use the full object if we only care about the `complete` part of the process.

Anyway, that's a lot of words and not a very concrete recommendation!  So, I'll suggest a code change in another comment.

Comment (ojeytonwilliams) [reply client/src/templates/Introduction/components/CertChallenge.js]: So, I figured it was easier to show than tell, and I pushed a commit.

The rationale for the name changes is pretty straightforward.  I don't know what `canClaim.status` or `canClaim.result` mean, but I can easily guess about `canClaimCert` and `certVerificationMessage`.  Well, I still don't know what the latter means, but it *is* a message about certification verification :man_shrugging: 

Same point for `userChecked` vs `verificationComplete`.  The former is a bit vague (what are we checking the user _for_). The latter is quite explicit and mirrors the patterns we usually use for ajax requests.  (pending, complete, etc. etc.).

**Review by ojeytonwilliams — APPROVED** [review]
The feature itself works rather nicely.  It's a nice improvement, thanks @Sembauke :purple_heart: 

That is to say, LGTM :+1:

**Comment by Sembauke** [comment]
No problem @ojeytonwilliams!

I agree that there are a lot of vague steps that have to be done in order to get this working properly.

But it's also a complicated process.

**Review by raisedadead — APPROVED** [review]
